From 260325dacc821f0a479ec1d7673ab5332159e993 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:10:52 -0400
Subject: [powerpc] tm: Fix GOT save offset for ABIv2

Message-id: <1410545655-205645-424-git-send-email-dzickus@redhat.com>
Patchwork-id: 94423
O-Subject: [RHEL7 PATCH 423/626] powerpc/tm: Fix GOT save offset for ABIv2
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 6403105bfda4d6934b39aeb85ff818b185b42de8
Author: Anton Blanchard <anton@samba.org>
Date:   Mon Mar 10 10:52:17 2014 +1100

    powerpc/tm: Fix GOT save offset for ABIv2

    The r2 TOC/GOT save offset is 40 on ABIv1 and 24 on ABIv2.

    Signed-off-by: Anton Blanchard <anton@samba.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/ppc_asm.h b/arch/powerpc/include/asm/ppc_asm.h
index 4fa163d..ca94485 100644
--- a/arch/powerpc/include/asm/ppc_asm.h
+++ b/arch/powerpc/include/asm/ppc_asm.h
@@ -191,8 +191,10 @@ END_FW_FTR_SECTION_IFSET(FW_FEATURE_SPLPAR)
 #define STK_REG(i)     __STK_REG(__REG_##i)
 
 #if defined(_CALL_ELF) && _CALL_ELF == 2
+#define STK_GOT  24
 #define __STK_PARAM(i) (32 + ((i)-3)*8)
 #else
+#define STK_GOT  40
 #define __STK_PARAM(i) (48 + ((i)-3)*8)
 #endif
 #define STK_PARAM(i) __STK_PARAM(__REG_##i)
diff --git a/arch/powerpc/kernel/tm.S b/arch/powerpc/kernel/tm.S
index 77206d7..ddf84c2 100644
--- a/arch/powerpc/kernel/tm.S
+++ b/arch/powerpc/kernel/tm.S
@@ -107,7 +107,7 @@ _GLOBAL(tm_reclaim)
  mflr r0
  stw r6, 8(r1)
  std r0, 16(r1)
- std r2, 40(r1)
+ std r2, STK_GOT(r1)
  stdu r1, -TM_FRAME_SIZE(r1)
 
  /* We've a struct pt_regs at [r1+STACK_FRAME_OVERHEAD]. */
@@ -288,7 +288,7 @@ dont_backup_fp:
  ld r0, 16(r1)
  mtcr r4
  mtlr r0
- ld r2, 40(r1)
+ ld r2, STK_GOT(r1)
 
  /* Load system default DSCR */
  ld r4, DSCR_DEFAULT@toc(r2)
@@ -311,7 +311,7 @@ _GLOBAL(__tm_recheckpoint)
  mflr r0
  stw r5, 8(r1)
  std r0, 16(r1)
- std r2, 40(r1)
+ std r2, STK_GOT(r1)
  stdu r1, -TM_FRAME_SIZE(r1)
 
  /* We've a struct pt_regs at [r1+STACK_FRAME_OVERHEAD].
@@ -447,7 +447,7 @@ restore_gprs:
  ld r0, 16(r1)
  mtcr r4
  mtlr r0
- ld r2, 40(r1)
+ ld r2, STK_GOT(r1)
 
  /* Load system default DSCR */
  ld r4, DSCR_DEFAULT@toc(r2)
-- 
1.7.1